engine/runtime-parser/src/main/java/org/enso/compiler/core/Persistance.java

```diff
@@ -0,0 +1,394 @@
+package org.enso.compiler.core;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+import org.enso.compiler.core.Persistance.Reference;
+
+public abstract class Persistance<T> {
+  private final Class<T> clazz;
+  private final boolean includingSubclasses;
+  private final int id;
+
+  protected Persistance(Class<T> clazz, boolean includingSubclasses, int id) {
+    this.clazz = clazz;
+    this.includingSubclasses = includingSubclasses;
+    this.id = id;
+  }
+
+  protected abstract void writeObject(T obj, Output out) throws IOException;
+  protected abstract T readObject(Input in) throws IOException, ClassNotFoundException;
+
+  public static interface Output extends DataOutput {
```

Comment:
This is the API for _suitably serializing_ `IR` tree. _"inline"_ means to store the object _"as it is"_ without any attempt to resolve subclasses or shared references.

`writeObject` stores a reference to an object.

_"inline"_ is useful for eliminating _"boxing"_ overhead when it is clear the stored object always has a single implementation. `Location` or `IdentifiedLocation` shall be stored _"inline"_.

`writeObject`/`readObject` shall be used to mitigate polymorphism. E.g.  `IR.Expression` & co. deserve to be stored with these methods.

Reply:
Worth documenting somewhere? (Fine if it is done in inline comments)

Reply:
Done in https://github.com/enso-org/enso/pull/8207#discussion_r1398105136